plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/AccountsPanel.java

```diff
@@ -0,0 +1,162 @@
+    FontUtil.convertFontToBold(messageLabel);
+
+    Composite activeAccountIndenter = new Composite(container, SWT.NONE);
+    GridLayoutFactory.swtDefaults().margins(EMAIL_LEFT_MARGIN, 0).applyTo(activeAccountIndenter);
+
+    Account activeAccount = loginService.getActiveAccount();
+    Label accountLabel = new Label(activeAccountIndenter, SWT.NONE);
+    accountLabel.setText(activeAccount.getEmail());
+    FontUtil.convertFontToBold(accountLabel);
+
+    Set<Account> accounts = loginService.listAccounts();
+    if (accounts.size() > 1) {
+      new Label(container, SWT.NONE).setText(Messages.LABEL_OTHER_ACCOUNTS);
+
+      Composite accountsIndenter = new Composite(container, SWT.NONE);
+      GridLayoutFactory.swtDefaults().margins(EMAIL_LEFT_MARGIN, 0).applyTo(accountsIndenter);
+
+      for (Account account : accounts) {
+        if (!account.getEmail().equals(activeAccount.getEmail())) {
+          Link link = new Link(accountsIndenter, SWT.NO_FOCUS);
+          link.setText("<a href=\"" + account.getEmail() + "\">" + account.getEmail() + "</a>"); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+          link.addSelectionListener(new SwitchAccountOnClick());
+        }
+      }
+    }
+  }
+
+  private void createButtons(Composite container) {
+    Composite buttonArea = new Composite(container, SWT.NONE);
+    GridDataFactory.defaultsFor(buttonArea).align(SWT.END, SWT.BEGINNING).applyTo(buttonArea);
+    GridLayoutFactory.swtDefaults().numColumns(2).margins(0, 0).applyTo(buttonArea);
+
+    Button addAccountButton = new Button(buttonArea, SWT.PUSH);
+    addAccountButton.setText(Messages.BUTTON_ACCOUNTS_PANEL_ADD_ACCOUNT);
+    addAccountButton.addSelectionListener(new LogInOnClick());
+    GridDataFactory.defaultsFor(addAccountButton).applyTo(addAccountButton);
+
+    if (loginService.isLoggedIn()) {
+      Button logOutButton = new Button(buttonArea, SWT.PUSH);
+      logOutButton.setText(Messages.BUTTON_ACCOUNTS_PANEL_LOGOUT);
+      logOutButton.addSelectionListener(new LogOutOnClick());
+      GridDataFactory.defaultsFor(logOutButton).applyTo(logOutButton);
+    }
+  }
+
+  private class LogInOnClick extends SelectionAdapter {
+    @Override
+    public void widgetSelected(SelectionEvent event) {
+      close();
+      loginService.logIn(null /* no custom dialog title */);
+    }
+  };
+
+  private class LogOutOnClick extends SelectionAdapter {
+    @Override
+    public void widgetSelected(SelectionEvent event) {
+      if (MessageDialog.openConfirm(getShell(),
+          Messages.LOGOUT_CONFIRM_DIALOG_TITILE, Messages.LOGOUT_CONFIRM_DIALOG_MESSAGE)) {
+        close();
+        loginService.logOutAll();
+      }
+    }
+  };
+
+  private class SwitchAccountOnClick extends SelectionAdapter {
+    @Override
+    public void widgetSelected(SelectionEvent event) {
+      String email = event.text;
+      loginService.switchActiveAccount(email);
+
+      close();
+      new AccountsPanel(getParentShell(), loginService).open();
```

Comment:
My concern was around resource consumption and the slightly increased GC-ing. Probably not too much for this rather simple dialog.